rbx_types/src/shared_string.rs

```diff
@@ -104,18 +102,25 @@ impl Drop for SharedString {
         // If the reference we're about to drop is the very last reference to
         // the buffer, we'll be able to unwrap it and remove it from the
         // SharedString cache.
-        if Arc::into_inner(self.data.take().unwrap()).is_some() {
-            let mut cache = match STRING_CACHE.lock() {
-                Ok(v) => v,
-                Err(_) => {
-                    // If the lock is poisoned, we should just leave it
-                    // alone so that we don't accidentally double-panic.
-                    return;
-                }
-            };
 
-            cache.remove(&self.hash);
-        }
+        // If we have the only strong reference, then
+        // we can be sure this is the only copy.  The inner Arc
+        // is never exposed so there will not be any downgraded
+        // weak copies anywhere other than the string cache.
+        // Therefore, it is not possible for another thread
+        // to change the strong count in between this check
+        // and `self.data` being dropped at the end of the scope.
+        if Arc::strong_count(&self.data) != 1 {
+            return;
+        };
+
+        let Ok(mut cache) = STRING_CACHE.lock() else {
+            // If the lock is poisoned, we should just leave it
+            // alone so that we don't accidentally double-panic.
+            return;
+        };
+
```

Comment:
I didn't realize that the string cache was upgrading weak references.  Maybe this double-check band-aid would work and not suffer from the slowness of directly locking every time that would happen if the strong count check was simply moved here.  I don't understand why the Arc methods in the standard library seem so limited.

```suggestion

        if Arc::strong_count(&self.data) != 1 {
            return;
        };

```